File: LeetCode/p438.py
from typing import List


class Solution:
    def findAnagrams(self, s:str, p:str)->List[int]:
        if len(p) > len(s): return []

        pCount, sCount = {}, {}

        for i in range(len(p)):
            pCount[p[i]] = 1 + pCount.get(p[i], 0)
            sCount[s[i]] = 1 + sCount.get(s[i], 0)

        # check first one is anagram or not
        # res = [0] if sCount == pCount else []
        res = []
        if sCount == pCount:
            res = [0]

        l = 0

        for r in range(len(p), len(s)):
            sCount[s[r]] = 1 + sCount.get(s[r], 0)
            sCount[s[l]] -= 1

            # if the value is zero then remove that
            if sCount[s[l]] == 0:
                sCount.pop(s[l])

            # update pointer
            l += 1

            if pCount == sCount:
                res.append(l)

        return res

File: LeetCode-Java/p438.java
import java.util.*;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        if (p.length() > s.length()) return res;

        Map<Character, Integer> pCount = new HashMap<>();
        Map<Character, Integer> sCount = new HashMap<>();

        for (int i = 0; i < p.length(); i++) {
            pCount.put(p.charAt(i), pCount.getOrDefault(p.charAt(i), 0) + 1);
            sCount.put(s.charAt(i), sCount.getOrDefault(s.charAt(i), 0) + 1);
        }

        if (sCount.equals(pCount)) {
            res.add(0);
        }

        int l = 0;
        for (int r = p.length(); r < s.length(); r++) {
            sCount.put(s.charAt(r), sCount.getOrDefault(s.charAt(r), 0) + 1);
            sCount.put(s.charAt(l), sCount.get(s.charAt(l)) - 1);

            if (sCount.get(s.charAt(l)) == 0) {
                sCount.remove(s.charAt(l));
            }

            l++;

            if (sCount.equals(pCount)) {
                res.add(l);
            }
        }

        return res;
    }
}
